src/documentMetadataManager.ts

```diff
@@ -0,0 +1,96 @@
+import { Disposable, TextDocument, Uri, workspace } from "vscode";
+import { Logger } from "./logging";
+import { getResourceManager } from "./storage/resourceManager";
+import { UriMetadataMap } from "./storage/types";
+
+const logger = new Logger("documentMetadataManager");
+
+/** Manager for VS Code {@link TextDocument}s that tracks metadata across document lifecycle events */
+export class DocumentMetadataManager {
+  private resourceManager = getResourceManager();
+
+  disposables: Disposable[] = [];
+
+  private constructor() {
+    this.disposables.push(
+      workspace.onDidOpenTextDocument(this.handleDocumentOpen, this),
+      workspace.onDidSaveTextDocument(this.handleDocumentSave, this),
+      workspace.onDidCloseTextDocument(this.handleDocumentClose, this),
+    );
```

Comment:
TODO: handle file renames caught while the extension is active
(but we can't track renames that happened outside the view of VS Code and/or the extension and will show the default codelens when metadata tracking is lost)

Reply:
Punting this to a follow-up branch since this one is already massive.